Feature/#75 커리큘럼 관리 기능 구현

JJimini opened the pull request:
## #️⃣ 연관된 이슈

close: #75

## 📝 작업 내용

- 커리큘럼이 추가됩니다.
    - 존재하지 않았던 `id`에 대해 커리큘럼을 추가합니다.
    - 커리큘럼 추가 시 참여자 커리큘럼도 추가됩니다.
    - 추가되는 데이터의 `id`는 가장 마지막에 추가되어 옵니다. (`null` 불가)
        - db에 3번까지 저장되어있다면, 새로 추가될 데이터의 id는 4번으로 옴
- 커리큘럼이 수정됩니다.
    - `id`는 존재하되, `name`이 변경되었다면 `name`을 수정합니다.
    - `id`는 존재하되, `itemOrder`가 변경되었다면 `itemOrder`를 수정합니다.
- 커리큘럼이 삭제됩니다.
    - 삭제된 커리큘럼 항목에 있는 `id`에 대해 커리큘럼을 삭제합니다.
- 커리큘럼의 아이템 순서를 정렬합니다.
    - 모든 수정이 끝난 후 커리큘럼을 아이템 순서 기준 오름차순으로 정렬하여 아이템 순서를 다시 업데이트합니다.

## 💬 리뷰 요구사항(선택)
- 제 설명이 잘 이해가 안되신다면, `curriculumCommandServiceTest`와 노션의 **[커리큘럼 관리]api**를 참고하시면 이해가 쉬울 것 같습니다.
- 브랜치명과 pr제목이 다른 이유는 작업 중 **수정**이라는 단어보다 **관리**가 더 맞다고 생각해서 변경했기 때문입니다..

## ⏰ 예상했던 소요시간/실제 소요시간(선택)
3시간/6시간

JJimini committed:
feat: 커리큘럼 관리 기능 구현

JJimini committed:
test: 커리큘럼 관리 기능 test 구현

JJimini committed:
feat: itemOrder 유효성 코드 추가
- itemOrder는 중복이 불가합니다.
- itemOrder는 Integer 범위를 넘어갈 수 없습니다.

JJimini committed:
Merge branch 'develop' into Feature/#75-커리큘럼_수정_기능_구현
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

JJimini committed:
docs: 커리큘럼 관리 api 문서 작성

JJimini committed:
fix: 참여자 예외처리 추가

JJimini committed:
fix: 참여자 예외처리 추가2

amaran-th commented on src/docs/asciidoc/curriculumItem.adoc:
```diff
@@ -14,32 +14,10 @@ endif::[]
-
-.Request Fields
-include::{snippets}/curriculum-update/http-request.adoc[]
+include::{snippets}/curriculum-manage/http-request.adoc[]
```
request fields(파라미터 설명)가 빠진 것 같아요!

JJimini replied:
이 부분이 request fields만 추가하면 자꾸 오류가 나서 생략해두었는데, 해결해두겠습니다!

JJimini replied:
파라미터 설명만 추가하면 자꾸 오류가 떠서 일단 생략해두었습니다. 해결해두겠습니다!

amaran-th commented on src/main/java/doore/study/application/CurriculumItemCommandService.java:
```diff
@@ -29,54 +33,105 @@ public class CurriculumItemCommandService {
+
+        for (CurriculumItem item : curriculumItems) {
+            int itemOrder = item.getItemOrder();
+            if (!uniqueItemOrders.add(itemOrder)) {
```
이 조건식은 메서드분리를 해서 직관적인 이름을 붙여주면 좀 더 이해하기 쉬울 것 같아요!

02ggang9 replied:
메서드 분리하면 스트림으로도 깔끔하게 작성할 수 있을 것 같습니다~

amaran-th commented on src/main/java/doore/study/application/CurriculumItemCommandService.java:
```diff
@@ -29,54 +33,105 @@ public class CurriculumItemCommandService {
+    }
+
+    private boolean existsCurriculumItem(Long curriculumItemId) {
+        return curriculumItemRepository.existsById(curriculumItemId);
```
새로 추가된 커리큘럼은 null 값으로 들어오는 것 아니었나요?? null 체크로 수정해도 괜찮을 것 같아요!
혹시 아니었다면 죄송합니다!

JJimini replied:
회의에서는 null 값이 들어온다고 했었는데, null이 들어오면 test 하기가 불편해서 id는 가장 마지막 번호로 추가해서 달라고 프론트분께 전달드렸습니다! 그래서 존재하지 않는 id를 찾아서 생성하는 방식으로 작성되었습니다! 자세한 설명은 `작업내용-커리큘럼이 추가됩니다 `부분을 참고해주세요!

02ggang9 replied:
메서드 이름 앞에 is 붙이는건 어떨까용?.? 지금 메서드 이름도 좋아 보입니다~

amaran-th commented on src/main/java/doore/study/application/CurriculumItemCommandService.java:
```diff
@@ -29,54 +33,105 @@ public class CurriculumItemCommandService {
 
-        int deletedOrder = curriculumItem.getItemOrder();
-        curriculumItemRepository.delete(curriculumItem);
+            if (!existingItem.getName().equals(requestItem.getName())) {
```
이 조건식도 메서드로 분리해도 될 것 같아요! 지금은 개인적으로 한 눈에 알아보기엔 조금 긴 것 같아서요!

amaran-th replied:
정확히는 CurriculumItem 엔티티의 메서드로 넣어주면 좋을 것 같네요~!

amaran-th commented on src/main/java/doore/study/application/CurriculumItemCommandService.java:
```diff
@@ -29,54 +33,105 @@ public class CurriculumItemCommandService {
+    private void sortCurriculum() {
+        List<CurriculumItem> sortedCurriculum = curriculumItemRepository.findAll()
+                .stream()
+                .sorted(Comparator.comparingInt(CurriculumItem::getItemOrder))
```
애플리케이션 수준에서 정렬 처리를 해주는 것보단 `ORDER BY`를 활용해서 JPQL 쿼리를 작성하는 편이 (사소하지만)성능면에서 더 좋을 것 같아요~!

amaran-th commented on src/main/java/doore/study/exception/StudyExceptionType.java:
```diff
@@ -6,7 +6,8 @@
     NOT_FOUND_STUDY(HttpStatus.NOT_FOUND, "스터디를 찾을 수 없습니다."),
     NOT_FOUND_STATUS(HttpStatus.BAD_REQUEST, "존재하지 않는 스터디 상태입니다."),
-    INVALID_ENDDATE(HttpStatus.BAD_REQUEST, "시작일보다 빠른 날짜에 종료할 수 없습니다.");
+    INVALID_ENDDATE(HttpStatus.BAD_REQUEST, "시작일보다 빠른 날짜에 종료할 수 없습니다."),
```
enum 값에서 의미가 좀 더 분명하게 드러나도록 `ENDDATE_BEFORE_STARTDATE`로 짓는 것도 괜찮을 것 같습니다!

JJimini replied:
이 부분은 제가 작성한 부분이 아닌데, "`;`" 수정하면서 제 변경사항에 들어간 것 같네요! 의견 남겨주세요! @lcqff 
근데 제 기억 상 의미가 좀 더 직관적으로 보이도록 INVALID를 가장 앞에 적도록 수정했었던 것 같네요!

lcqff replied:
미나님이 작성하신 코드가 아니니 원하시면 수정하시고 괜찮으시면 패스하셔도 될 것 같습니다! 👀

shkisme replied:
Trailing commas (후행 쉼표) 표기의 이점을 찾아서 공유해드립니다~

```kotlin
class Person(
    val firstName: String,
    val lastName: String,
    val age: Int, // trailing comma
)
```
후행 쉼표를 사용하면 몇 가지 이점이 있습니다:

- **변경된 값에만 초점을 맞추기 때문에 git Diff를 더 깔끔하게 만들 수 있습니다.**
- 요소를 조작할 때 쉼표를 추가하거나 삭제할 필요가 없으므로 요소를 쉽게 추가하고 순서를 바꿀 수 있습니다.
- 예를 들어 객체 이니셜라이저의 경우 코드 생성이 간소화됩니다. 마지막 요소에도 쉼표를 넣을 수 있습니다.
- 후행 쉼표는 전적으로 선택 사항이며 쉼표가 없어도 코드가 계속 작동합니다. Kotlin 스타일 가이드에서는 선언 사이트에서 후행 쉼표 사용을 권장하며 호출 사이트에서는 사용자의 재량에 맡깁니다.

amaran-th reviewed (approved):
구현하시느라 수고 많으셨습니다~!
전체적으로 요구사항을 잘 구현해주신 것 같아요!
개인적으로 신경쓰이는 부분에 코멘트 남겨놓았는데 꼭 모두 반영하실 필요는 없습니다!
그런 의미에서 approve하겠습니다~

JJimini committed:
fix: controller test 삭제
- api docs test와 중복된 내용의 테스트로, 삭제하였습니다.

JJimini committed:
fix: controllerTest 삭제 (최종)

lcqff commented on src/main/java/doore/study/application/CurriculumItemCommandService.java:
```diff
@@ -29,54 +33,105 @@ public class CurriculumItemCommandService {
+            if (!uniqueItemOrders.add(itemOrder)) {
+                throw new CurriculumItemException(INVALID_ITEM_ORDER);
+            }
+            if (itemOrder < Integer.MIN_VALUE || itemOrder > Integer.MAX_VALUE) {
```
아직 논의 안된 부분이긴 한데 만들 수 있는 커리큘럼의 최대 개수도 생각해봐야겠네요... 주석으로 최대 개수 설정 필요하다고 todo 달아주실수 있을까요?
그리고 itemOrder가 음수가 될 일은 없을 것 같으니 Min Value는 0이하로 해도 괜찮지 않을까요?

shkisme replied:
클라이언트에서 음수로 보낼 경우 막아야 할 것 같긴 해용!
이런 처리는 Dto에서 @PostiveOrZero 사용하는거 어떤가 합니다~

JJimini replied:
클라이언트에서 음수로 와도 괜찮습니다! 순서만 맞다면, 중복되지 않음 & Integer 범위만 넘어가지 않는다면 정렬해서 다시 Order를 부여하기 때문에 문제가 되지 않을 것 같습니다! 
ex) -1 , 5, 3 -> -1, 3, 5 -> 1, 2, 3

이 코드는 의도적으로 Integer 범위를 이외의 값을 보내는 경우를 대비해서 작성한 코드입니다!

02ggang9 replied:
메서드 이름 checkItemOrderRange 처럼 동사가 앞으로 오는건 어떠실까용?.?

요것도 위의 리뷰와 마찬가지로 스트림으로 간단하게 표현할 수 있을 것 같습니다~ foreach가 더 가독성이 좋다고 생각하신다면 foreach 사용도 좋습니다~!

lcqff commented on src/main/java/doore/study/application/CurriculumItemCommandService.java:
```diff
@@ -29,54 +33,105 @@ public class CurriculumItemCommandService {
+    private void validItemOrderCheck(List<CurriculumItem> curriculumItems) {
+        Set<Integer> uniqueItemOrders = new HashSet<>();
+
+        for (CurriculumItem item : curriculumItems) {
```
itemOrder 같은 경우 프론트 쪽에서 List의 인덱스로 넘겨주는 걸로 알고 있는데 그래도 백엔드 쪽에서 order의 unique함 같은 잘못되면 치명적일수도 있는? 데이터의 유효성을 검사해야 하나요? 
이건 리뷰라기보단 질문입니다!

JJimini replied:
프론트에서 오는 값을 한 번 더 검증하는 과정이라고 보시면 됩니다. 백엔드에서는 마지막에 오름차순 정렬해서 다시 itemOrder를 부여한 후 db를 업데이트합니다. 프론트에서 itemOrder에 어떤 값이 오든지 크고 작음만 맞으면 정상 정렬되어 itemOrder가 부여되지만, 중복된 값이 오는 경우 정렬에 문제가 생기기 때문에 한 번 더 검사하는 겁니다!

lcqff commented on src/main/java/doore/study/application/CurriculumItemCommandService.java:
```diff
@@ -29,54 +33,105 @@ public class CurriculumItemCommandService {
+    private void createCurriculum(Long studyId, List<CurriculumItem> curriculumItems) {
+        curriculumItems.stream()
+                .filter(curriculumItem -> !existsCurriculumItem(curriculumItem.getId()))
+                .forEach(curriculumItem -> createCurriculumItemAndAssignToParticipants(studyId, curriculumItem));
```
👍👍👍

lcqff commented on src/main/java/doore/study/application/CurriculumItemCommandService.java:
```diff
@@ -29,54 +33,105 @@ public class CurriculumItemCommandService {
-        if (participantCurriculumItem.getIsChecked().equals(false)) participantCurriculumItem.complete();
-        else participantCurriculumItem.incomplete();
-    }
+    private void sortCurriculum() {
```
연속적인 오름차순이 되지 않을 경우도 존재하나요??

JJimini replied:
네넵! 커리큘럼이 삭제된 경우 오름차순 정렬하였을 때 불연속적입니다! 그래서 다시 itemOrder를 재부여하는 과정을 거칩니다!

ex) id : 2번이 삭제된 경우 (itemOrder는 순서를 수정한 후 itemOrder 라고 가정), 
id : 1, itemOrder : 3
id : 3, itemOrder : 4
id : 4, itemOrder : 1

itemOrder 오름차순 정렬 -> 1, 3, 4 (불연속적)
itemOrder 재부여 -> 1, 2, 3

최종적으로 db에 저장된 형태
id : 1, itemOrder : 2
id : 3, itemOrder : 3
id : 4, itemOrder : 1

serviceTest에 이 부분에 대한 코드가 구현되어있습니다~! 제 설명이 부족해서.. 혹시 이해가 안되신다면 코드 참조해주세용

lcqff replied:
프론트에서 순서 정렬해서 주는게 아니군용~~~!! 확인입니다😄

lcqff reviewed (commented):
리뷰 늦어서 죄송합니다!!!! 여행갔다가 바로 친가 외가 가고 기절해있었습니다ㅠㅠ 늙었나봅니다
사소한 질문이랑 요청 몇개 남겨놨습니다~!

shkisme commented on src/main/java/doore/study/domain/CurriculumItem.java:
```diff
@@ -45,18 +45,19 @@ public class CurriculumItem extends BaseEntity {
 
     @Builder
-    private CurriculumItem(String name, Integer itemOrder, Study study) {
+    private CurriculumItem(Long id, String name, Integer itemOrder, Study study) {
+        this.id = id;
```
id를 직접 설정해주신 이유가 궁금합니다~!

JJimini replied:
이 부분은 테스트에서 테스트 데이터 만든다고 적었던 것 같아요!

shkisme commented on src/main/java/doore/study/application/CurriculumItemCommandService.java:
```diff
@@ -29,54 +33,105 @@ public class CurriculumItemCommandService {
+        for (int i = 0; i < sortedCurriculum.size(); i++) {
+            CurriculumItem item = sortedCurriculum.get(i);
+            item.updateItemOrder(i + 1);
+        }
```
```suggestion
         IntStream.range(0, sortedCurriculum.size())
                .forEach(i -> sortedCurriculum.get(i).updateItemOrder(i + 1));
```
이렇게 작성할 수 있겠네요~ Nit입니다~! `IntStream` 가독성이 별로인 것 같다면 반영 안하셔도 되요!

shkisme reviewed (approved):
복잡한 요구사항이었던 것 같은데 잘 구현해주신 것 같습니다~~!! 👍👍👍

JJimini committed:
refactor : itemOrder 유효성 검사 메서드 분리

JJimini committed:
refactor : name 및 itemOrder 업데이트 관련 메서드 분리

JJimini committed:
refactor : 정렬 방법 변경

JJimini committed:
refactor : itemOrder 재부여 코드를 IntStream을 활용한 방법으로 변경

JJimini committed:
Merge branch 'develop' into Feature/#75-커리큘럼_수정_기능_구현

# Conflicts:
#	src/test/java/doore/restdocs/RestDocsTest.java

02ggang9 commented on src/main/java/doore/study/api/CurriculumItemController.java:
```diff
@@ -1,46 +1,32 @@
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RestController;
 
+@Validated
```
깨알 디테일 굳굳 입니다~

02ggang9 commented on src/main/java/doore/study/application/CurriculumItemCommandService.java:
```diff
@@ -29,54 +33,102 @@ public class CurriculumItemCommandService {
-        CurriculumItem curriculumItem = curriculumItemRepository.findById(curriculumId)
-                .orElseThrow(() -> new CurriculumItemException(NOT_FOUND_CURRICULUM_ITEM));
-        curriculumItem.update(request.name());
+    private void deleteCurriculum(List<CurriculumItem> deletedCurriculumItems) {
+        for (CurriculumItem requestItem : deletedCurriculumItems) {
+            curriculumItemRepository.deleteById(requestItem.getId());
+        }
```
요 부분도 스트림으로 처리할 수 있을 것 같습니다~ 

제가 코드 블럭이 딱 맞는 스트림 형식을 좋아해서.. foreach를 선호하시면 foreach도 좋습니다~! 스트림이 성능이 더 떨어지는 것으로 알고 있습니다..ㅎ

02ggang9 reviewed (approved):
코드 작성하신다고 수고 많으셨습니다~ 코드 깔끔하시네용 굳굳입니다!

JJimini committed:
refactor: 리뷰 수정사항 수정

JJimini committed:
fix: 커리큘럼 체크 오류 수정
- 커리큘럼 url에서 participantId를 받아옵니다.
- 참여자가 직접 체크하는 방식이므로 참여자Id를 통해 정보를 찾아와 변경합니다.

JJimini committed:
test: url 수정에 따른 테스트 수정

JJimini committed:
test: 커리큘럼 상태 변경 테스트 코드 추가

JJimini committed:
docs: 커리큘럼 관리 관련 문서 추가

JJimini committed:
fix: 중복된 테스트 요소 삭제

JJimini committed:
fix: 커리큘럼 상태 변경 시 참여자아이디와 커리큘럼 아이디를 모두 활용하는 방식으로 변경
